File: cpp/src/enrich.cpp
// SPDX-FileCopyrightText: 2025 Nicolas Pelletier
// SPDX-License-Identifier: BSD-2-Clause
// Enrichment logic: formula pretty-printer, signal collector, diagnostics.
#include <aletheia/enrich.hpp>

#include <aletheia/detail/rational_renderer.hpp>

#include <algorithm>
#include <cstdint>
#include <format>
#include <string>
#include <string_view>
#include <type_traits>
#include <utility>
#include <variant>
#include <vector>

namespace aletheia {

// Greek capital delta, the change-predicate prefix.
constexpr std::string_view k_delta = "\u0394";

static auto format_timebound(Timestamp t) -> std::string {
    auto us = t.count();
    if (us % us_per_second == 0)
        return std::format("{}s ", us / us_per_second);
    if (us % us_per_millisecond == 0)
        return std::format("{}ms ", us / us_per_millisecond);
    return std::format("{}\u03bcs ", us);
}

// The token a value-comparison predicate renders between its signal and its
// threshold.  The five such predicates differ in nothing else.
template<typename T>
static constexpr auto comparison_token() -> std::string_view {
    if constexpr (std::is_same_v<T, Equals>)
        return "=";
    else if constexpr (std::is_same_v<T, LessThan>)
        return "<";
    else if constexpr (std::is_same_v<T, GreaterThan>)
        return ">";
    else if constexpr (std::is_same_v<T, LessThanOrEqual>)
        return "<=";
    else if constexpr (std::is_same_v<T, GreaterThanOrEqual>)
        return ">=";
    else
        static_assert(sizeof(T) == 0, "not a value-comparison predicate");
}

// Every predicate is one of four shapes, named below by its members rather
// than by its type; a predicate of a new shape fails the final static_assert.
static auto format_predicate(const Predicate& p) -> std::string {
    return std::visit(
        [](const auto& v) -> std::string {
            using T = std::decay_t<decltype(v)>;
            if constexpr (requires { v.value; })
                return std::format("{} {} {}", std::string_view{v.signal}, comparison_token<T>(),
                                   detail::format_rational(v.value.get()));
            else if constexpr (requires {
                                   v.min;
                                   v.max;
                               })
                return std::format("{} <= {} <= {}", detail::format_rational(v.min.get()),
                                   std::string_view{v.signal},
                                   detail::format_rational(v.max.get()));
            else if constexpr (requires { v.delta; })
                // The sign of the delta says which direction the change bounds.
                return std::format("{}{} {} {}", k_delta, std::string_view{v.signal},
                                   v.delta.get() >= Rational{0, 1} ? ">=" : "<=",
                                   detail::format_rational(v.delta.get()));
            else if constexpr (requires { v.tolerance; })
                return std::format("|{}{}| <= {}", k_delta, std::string_view{v.signal},
                                   detail::format_rational(v.tolerance.get()));
            else
                static_assert(sizeof(T) == 0, "Unhandled predicate shape in format_predicate");
        },
        p);
}

static auto predicate_signal(const Predicate& p) -> SignalName {
    return std::visit([](const auto& v) -> SignalName { return v.signal; }, p);
}

// Walks the tree by shape (a predicate, two children, or one), so an
// alternative of an existing shape needs no branch here.
static void collect_signals_into(const LtlFormula& f, std::vector<SignalName>& signals) {
    f.visit([&signals](const auto& v) {
        using T = std::decay_t<decltype(v)>;
        if constexpr (requires { v.predicate; }) {
            auto name = predicate_signal(v.predicate);
            if (!std::ranges::contains(signals, name))
                signals.push_back(name);
        } else if constexpr (requires {
                                 v.left;
                                 v.right;
                             }) {
            collect_signals_into(*v.left, signals);
            collect_signals_into(*v.right, signals);
        } else if constexpr (requires { v.formula; }) {
            collect_signals_into(*v.formula, signals);
        } else {
            static_assert(sizeof(T) == 0, "Unhandled formula shape in collect_signals_into");
        }
    });
}

static auto format_formula_inner(const LtlFormula& f, bool parenthesize_binary) -> std::string;

static auto wrap_if_binary(std::string s, bool parenthesize) -> std::string {
    return parenthesize ? "(" + std::move(s) + ")" : std::move(s);
}

template<typename Node>
static auto format_binary(const Node& v, std::string_view op, bool parenthesize) -> std::string {
    return wrap_if_binary(format_formula_inner(*v.left, true) + " " + std::string{op} + " " +
                              format_formula_inner(*v.right, true),
                          parenthesize);
}

template<typename Node>
static auto format_metric_binary(const Node& v, std::string_view op, bool parenthesize)
    -> std::string {
    return wrap_if_binary(format_formula_inner(*v.left, true) + " " + std::string{op} + " within " +
                              format_timebound(v.bound) + format_formula_inner(*v.right, true),
                          parenthesize);
}

// Detect Never pattern: Always{Not{Atomic{p}}} — returns empty string if not.
static auto try_format_never(const Always& v) -> std::string {
    if (auto* n = std::get_if<Not>(&v.formula->value))
        if (auto* a = std::get_if<Atomic>(&n->formula->value))
            return "never " + format_predicate(a->predicate);
    return {};
}

// Inner formatter: parenthesize_binary wraps binary operators in parens when
// they appear as children of other binary operators.  The whole rendering is
// byte-identical to the Python and Go formatters; a probe under probes/
// compares this formatter against Python's over every alternative.
static auto format_formula_inner(const LtlFormula& f, bool parenthesize_binary) -> std::string {
    return f.visit([parenthesize_binary](const auto& v) -> std::string {
        using T = std::decay_t<decltype(v)>;
        if constexpr (std::is_same_v<T, Atomic>) {
            return format_predicate(v.predicate);
        } else if constexpr (std::is_same_v<T, Not>) {
            return "not(" + format_formula_inner(*v.formula, false) + ")";
        } else if constexpr (std::is_same_v<T, And>) {
            return format_binary(v, "and", parenthesize_binary);
        } else if constexpr (std::is_same_v<T, Or>) {
            return format_binary(v, "or", parenthesize_binary);
        } else if constexpr (std::is_same_v<T, Next>) {
            return "next(" + format_formula_inner(*v.formula, false) + ")";
        } else if constexpr (std::is_same_v<T, WeakNext>) {
            return "weak_next(" + format_formula_inner(*v.formula, false) + ")";
        } else if constexpr (std::is_same_v<T, Always>) {
            auto never = try_format_never(v);
            return never.empty() ? "always(" + format_formula_inner(*v.formula, false) + ")"
                                 : never;
        } else if constexpr (std::is_same_v<T, Eventually>) {
            return "eventually(" + format_formula_inner(*v.formula, false) + ")";
        } else if constexpr (std::is_same_v<T, Until>) {
            return format_binary(v, "until", parenthesize_binary);
        } else if constexpr (std::is_same_v<T, Release>) {
            return format_binary(v, "release", parenthesize_binary);
        } else if constexpr (std::is_same_v<T, MetricAlways>) {
            return "always within " + format_timebound(v.bound) + "(" +
                   format_formula_inner(*v.formula, false) + ")";
        } else if constexpr (std::is_same_v<T, MetricEventually>) {
            return "eventually within " + format_timebound(v.bound) + "(" +
                   format_formula_inner(*v.formula, false) + ")";
        } else if constexpr (std::is_same_v<T, MetricUntil>) {
            return format_metric_binary(v, "until", parenthesize_binary);
        } else if constexpr (std::is_same_v<T, MetricRelease>) {
            return format_metric_binary(v, "release", parenthesize_binary);
        } else {
            static_assert(sizeof(T) == 0, "Unhandled formula type in format_formula");
        }
    });
}

auto format_formula(const LtlFormula& f) -> std::string {
    return format_formula_inner(f, false);
}

auto collect_signals(const LtlFormula& f) -> std::vector<SignalName> {
    std::vector<SignalName> signals;
    collect_signals_into(f, signals);
    return signals;
}

auto build_diagnostic(const LtlFormula& f) -> PropertyDiagnostic {
    return PropertyDiagnostic{
        .signals = collect_signals(f),
        .formula_desc = format_formula(f),
    };
}

} // namespace aletheia

File: rust/src/enrich.rs
// SPDX-FileCopyrightText: 2025 Nicolas Pelletier
// SPDX-License-Identifier: BSD-2-Clause

//! Client-side violation enrichment (Slice R4b).
//!
//! The verified core reports a violation with a raw `reason` string but no
//! signal context. This module reconstructs that context **client-side** — the
//! same approach as the Go (`enrich.go`) and Python (`_enrichment.py`) bindings:
//! from each registered [`Formula`] it derives a [`PropertyDiagnostic`] (the
//! signals the formula references + a human-readable description), and on a
//! violation it pairs those signals with their last-known values to build an
//! [`Enrichment`](crate::Enrichment).
//!
//! Rational values render through the verified kernel's `formatℚ` (via
//! [`format_rational`]) — the same renderer the
//! check DSL's `condition_desc` now uses, so a threshold and a signal value render
//! identically across both surfaces *and* byte-identically to the Python / Go /
//! C++ bindings (the single source of truth lives in Agda; there is no local
//! fallback). `enriched_reason` is a human-readable diagnostic; no gate pins it to
//! the peers' byte-for-byte output (their tests assert substrings, not equality).

use crate::backend::format_rational;
use crate::error::Error;
use crate::ltl::{Formula, Predicate};
use crate::types::{Rational, TimeBound};

/// Diagnostic metadata derived from a formula, cached per registered property.
#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct PropertyDiagnostic {
    /// Signals referenced by the formula, deduplicated in first-seen order.
    pub signals: Vec<String>,
    /// Human-readable rendering of the formula.
    pub formula_desc: String,
}

/// Build the diagnostic for a formula.
///
/// # Errors
/// Propagates an [`Error`] if the kernel rational renderer's library cannot be
/// loaded (see [`format_rational`]); callers render thresholds at property-setup
/// time, where a load failure is a legitimate setup error.
pub(crate) fn build_diagnostic(f: &Formula) -> Result<PropertyDiagnostic, Error> {
    Ok(PropertyDiagnostic {
        signals: collect_signals(f),
        formula_desc: format_formula(f)?,
    })
}

// ── Signal collection ─────────────────────────────────────────────────────────

/// The signal a predicate concerns.
fn predicate_signal(p: &Predicate) -> &str {
    match p {
        Predicate::Equals { signal, .. }
        | Predicate::LessThan { signal, .. }
        | Predicate::GreaterThan { signal, .. }
        | Predicate::LessThanOrEqual { signal, .. }
        | Predicate::GreaterThanOrEqual { signal, .. }
        | Predicate::Between { signal, .. }
        | Predicate::ChangedBy { signal, .. }
        | Predicate::StableWithin { signal, .. } => signal,
    }
}

/// All signals referenced in a formula, deduplicated in first-seen order.
fn collect_signals(f: &Formula) -> Vec<String> {
    let mut out = Vec::new();
    collect_into(f, &mut out);
    out
}

fn collect_into(f: &Formula, out: &mut Vec<String>) {
    match f {
        Formula::Atomic(p) => {
            let s = predicate_signal(p);
            if !out.iter().any(|x| x == s) {
                out.push(s.to_string());
            }
        }
        Formula::Not(inner)
        | Formula::Next(inner)
        | Formula::WeakNext(inner)
        | Formula::Always(inner)
        | Formula::Eventually(inner)
        | Formula::MetricAlways { inner, .. }
        | Formula::MetricEventually { inner, .. } => collect_into(inner, out),
        Formula::And(l, r)
        | Formula::Or(l, r)
        | Formula::Until(l, r)
        | Formula::Release(l, r)
        | Formula::MetricUntil {
            left: l, right: r, ..
        }
        | Formula::MetricRelease {
            left: l, right: r, ..
        } => {
            collect_into(l, out);
            collect_into(r, out);
        }
    }
}

// ── Formatting ────────────────────────────────────────────────────────────────

fn format_predicate(p: &Predicate) -> Result<String, Error> {
    Ok(match p {
        Predicate::Equals { signal, value } => format!("{signal} = {}", format_rational(*value)?),
        Predicate::LessThan { signal, value } => format!("{signal} < {}", format_rational(*value)?),
        Predicate::GreaterThan { signal, value } => {
            format!("{signal} > {}", format_rational(*value)?)
        }
        Predicate::LessThanOrEqual { signal, value } => {
            format!("{signal} <= {}", format_rational(*value)?)
        }
        Predicate::GreaterThanOrEqual { signal, value } => {
            format!("{signal} >= {}", format_rational(*value)?)
        }
        Predicate::Between { signal, min, max } => {
            format!(
                "{} <= {signal} <= {}",
                format_rational(*min)?,
                format_rational(*max)?
            )
        }
        Predicate::ChangedBy { signal, delta } => {
            if delta.numerator() >= 0 {
                format!("Δ{signal} >= {}", format_rational(*delta)?)
            } else {
                format!("Δ{signal} <= {}", format_rational(*delta)?)
            }
        }
        Predicate::StableWithin { signal, tolerance } => {
            format!("|Δ{signal}| <= {}", format_rational(*tolerance)?)
        }
    })
}

const US_PER_SECOND: u64 = 1_000_000;
const US_PER_MILLISECOND: u64 = 1_000;

fn format_timebound(t: TimeBound) -> String {
    let us = t.micros();
    if us.is_multiple_of(US_PER_SECOND) {
        format!("{}s", us / US_PER_SECOND)
    } else if us.is_multiple_of(US_PER_MILLISECOND) {
        format!("{}ms", us / US_PER_MILLISECOND)
    } else {
        format!("{us}μs")
    }
}

/// Human-readable rendering of a formula (the `never P` shorthand mirrors the peers).
fn format_formula(f: &Formula) -> Result<String, Error> {
    format_inner(f, false)
}

/// `parenthesize` wraps binary operators when nested in another binary operator.
fn format_inner(f: &Formula, parenthesize: bool) -> Result<String, Error> {
    let wrap = |s: String| if parenthesize { format!("({s})") } else { s };
    Ok(match f {
        Formula::Atomic(p) => format_predicate(p)?,
        Formula::Not(inner) => format!("not({})", format_inner(inner, false)?),
        Formula::And(l, r) => wrap(format!(
            "{} and {}",
            format_inner(l, true)?,
            format_inner(r, true)?
        )),
        Formula::Or(l, r) => wrap(format!(
            "{} or {}",
            format_inner(l, true)?,
            format_inner(r, true)?
        )),
        Formula::Next(inner) => format!("next({})", format_inner(inner, false)?),
        Formula::WeakNext(inner) => format!("weak_next({})", format_inner(inner, false)?),
        Formula::Always(inner) => {
            // Detect the `never P` shape: Always(Not(Atomic(p))).
            if let Formula::Not(n) = inner.as_ref() {
                if let Formula::Atomic(p) = n.as_ref() {
                    return Ok(format!("never {}", format_predicate(p)?));
                }
            }
            format!("always({})", format_inner(inner, false)?)
        }
        Formula::Eventually(inner) => format!("eventually({})", format_inner(inner, false)?),
        Formula::Until(l, r) => wrap(format!(
            "{} until {}",
            format_inner(l, true)?,
            format_inner(r, true)?
        )),
        Formula::Release(l, r) => wrap(format!(
            "{} release {}",
            format_inner(l, true)?,
            format_inner(r, true)?
        )),
        Formula::MetricAlways { bound, inner } => {
            format!(
                "always within {} ({})",
                format_timebound(*bound),
                format_inner(inner, false)?
            )
        }
        Formula::MetricEventually { bound, inner } => {
            format!(
                "eventually within {} ({})",
                format_timebound(*bound),
                format_inner(inner, false)?
            )
        }
        Formula::MetricUntil { bound, left, right } => wrap(format!(
            "{} until within {} {}",
            format_inner(left, true)?,
            format_timebound(*bound),
            format_inner(right, true)?
        )),
        Formula::MetricRelease { bound, left, right } => wrap(format!(
            "{} release within {} {}",
            format_inner(left, true)?,
            format_timebound(*bound),
            format_inner(right, true)?
        )),
    })
}

/// Append the raw core reason to a base reason, if present.
fn with_core(base: String, core_reason: &str) -> String {
    if core_reason.is_empty() {
        base
    } else {
        format!("{base} [core: {core_reason}]")
    }
}

/// The enriched reason when no signal values are available — the formula
/// description plus the raw core reason. Renders no rationals, so it is infallible
/// and doubles as the graceful degrade target if observed-value rendering ever
/// fails (unreachable: the values came from an extraction that already loaded the
/// library, so the renderer's library is loaded too).
pub(crate) fn reason_without_values(diag: &PropertyDiagnostic, core_reason: &str) -> String {
    with_core(format!("violated: {}", diag.formula_desc), core_reason)
}

/// Build the enriched reason string from a diagnostic, the values extracted for
/// its signals, and the raw core reason. Mirrors the Go / Python formatters.
///
/// # Errors
/// Propagates an [`Error`] if the kernel rational renderer's library cannot be
/// loaded while rendering an observed value (see [`format_rational`]). Callers on
/// the eval path degrade to [`reason_without_values`] rather than fail an
/// already-processed frame.
pub(crate) fn format_enriched_reason(
    diag: &PropertyDiagnostic,
    values: &[(String, Rational)],
    core_reason: &str,
) -> Result<String, Error> {
    let mut parts: Vec<String> = Vec::new();
    for sig in &diag.signals {
        if let Some((_, v)) = values.iter().find(|(n, _)| n == sig) {
            parts.push(format!("{sig} = {}", format_rational(*v)?));
        }
    }
    if parts.is_empty() {
        return Ok(reason_without_values(diag, core_reason));
    }
    Ok(with_core(
        format!("{} (formula: {})", parts.join(", "), diag.formula_desc),
        core_reason,
    ))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::check;

    fn r(n: i64) -> Rational {
        Rational::integer(n)
    }

    #[test]
    fn never_shape_and_signals() {
        let f = check::signal("Speed").never_exceeds(220);
        let d = build_diagnostic(f.formula()).unwrap();
        assert_eq!(d.signals, vec!["Speed".to_string()]);
        assert_eq!(d.formula_desc, "always(Speed <= 220)");
    }

    #[test]
    fn never_equals_renders_as_never() {
        // never_equals builds Always(Not(Atomic(Equals))) → the `never P` shape.
        let f = check::signal("Mode").never_equals(3);
        assert_eq!(
            build_diagnostic(f.formula()).unwrap().formula_desc,
            "never Mode = 3"
        );
    }

    #[test]
    fn collect_dedups_across_when_then() {
        let f = check::when("Brake")
            .exceeds(50)
            .then("Light")
            .equals(1)
            .within(100)
            .unwrap();
        let d = build_diagnostic(f.formula()).unwrap();
        // Both signals referenced, in first-seen order; metric bound rendered.
        assert_eq!(d.signals, vec!["Brake".to_string(), "Light".to_string()]);
        assert!(
            d.formula_desc.contains("within 100ms"),
            "got: {}",
            d.formula_desc
        );
    }

    #[test]
    fn enriched_reason_with_and_without_values() {
        let diag = PropertyDiagnostic {
            signals: vec!["Speed".to_string()],
            formula_desc: "always(Speed < 220)".to_string(),
        };
        let with = format_enriched_reason(&diag, &[("Speed".to_string(), r(245))], "").unwrap();
        assert_eq!(with, "Speed = 245 (formula: always(Speed < 220))");
        let without = format_enriched_reason(&diag, &[], "window expired").unwrap();
        assert_eq!(
            without,
            "violated: always(Speed < 220) [core: window expired]"
        );
    }
}
